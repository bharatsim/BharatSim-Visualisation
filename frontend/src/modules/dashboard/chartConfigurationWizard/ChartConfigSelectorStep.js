import React from 'react';
import PropTypes from 'prop-types';
import { useForm as useReactForm } from 'react-hook-form';
import {
  Box,
  Button,
  Tabs,
  Typography,
  Divider,
  Tab,
  makeStyles,
  TextField,
} from '@material-ui/core';

import DatasourceSelector from '../dashboardConfigSelector/DatasourceSelector';
import ConfigSelector from '../dashboardConfigSelector/ConfigSelector';
import ButtonGroup from '../../../uiComponent/ButtonGroup';
import { useFooterStyles } from './styles';

const DATASOURCE_SELECTOR_KEY = 'dataSource';
const CHART_NAME_KEY = 'chartName';

const useStyles = makeStyles((theme) => ({
  container: {
    height: `calc(100vh - ${theme.spacing(69)}px)`,
    overflowY: 'scroll',
  },
  tabContainer: {
    padding: theme.spacing(0, 8),
    backgroundColor: theme.colors.primaryColorScale['50'],
    borderStyle: 'solid',
    borderColor: theme.colors.grayScale['200'],
    borderWidth: '1px 0 1px 0',
    margin: theme.spacing(2, -6, 8, -6),
  },
  tabRoot: {
    height: theme.spacing(12),
  },
  textFieldRoot: {
    minWidth: theme.spacing(80),
  },
}));

function ChartConfigSelectorStep({ existingConfig, chartType, onApply, backToChartType }) {
  const footerClasses = useFooterStyles();
  const [selectedTab] = React.useState(0);
  const classes = useStyles();
  const chartName = existingConfig[CHART_NAME_KEY] || 'Untitled Chart';
  const reactForm = useReactForm({ mode: 'onChange' });
  const {
    register,
    handleSubmit,
    errors,
    getValues,
    control,
    watch,
    setValue,
    formState,
  } = reactForm;
  const watchShowOtherConfig = watch(DATASOURCE_SELECTOR_KEY, undefined);

  function isEditMode() {
    return Object.keys(existingConfig).length !== 0;
  }

  function onApplyClick(data) {
    onApply(data);
  }

  function resetValue(keys) {
    keys.forEach((key) => setValue(key, undefined));
  }

  return (
    <Box>
      <form onSubmit={handleSubmit(onApplyClick)}>
        <Box className={classes.tabContainer}>
          <Tabs value={selectedTab} indicatorColor="primary" aria-label="disabled tabs example">
            <Tab label="Data" classes={{ root: classes.tabRoot }} />
          </Tabs>
        </Box>
        <Box className={classes.container}>
          <Box px={2} pb={6}>
            <Box mb={2}>
              <Typography variant="subtitle2">Chart Name</Typography>
            </Box>
            <TextField
              id={CHART_NAME_KEY}
              data-testid={CHART_NAME_KEY}
              label="Add chart name"
              variant="filled"
              classes={{ root: classes.textFieldRoot }}
              inputProps={{ name: CHART_NAME_KEY, defaultValue: chartName }}
              inputRef={register({ required: true })}
            />
          </Box>
          <Divider />
          <Box px={2} py={6}>
            <DatasourceSelector
              disabled={isEditMode()}
              control={control}
              name={DATASOURCE_SELECTOR_KEY}
              error={errors[DATASOURCE_SELECTOR_KEY]}
              header="Data Source"
              id="dropdown-dataSources"
              label="select data source"
            />
          </Box>
          {watchShowOtherConfig && (
            <>
              <Divider />
              <Box pt={6}>
                <ConfigSelector
                  dataSourceId={getValues(DATASOURCE_SELECTOR_KEY)}
                  isEditMode={isEditMode()}
                  errors={errors}
                  chartType={chartType}
                  register={register}
                  control={control}
                  watch={watch}
                  resetValue={resetValue}
                />
              </Box>
            </>
          )}
          <Box className={footerClasses.footer}>
            <Divider />
            <Box mt={3} display="flex" justifyContent="flex-end">
              <ButtonGroup>
                {!isEditMode() && (
                  <Button variant="text" size="small" onClick={backToChartType}>
                    Back to chart type
                  </Button>
                )}
                <Button
                  type="submit"
                  variant="contained"
                  color="primary"
                  size="small"
                  disabled={!formState.isValid}
                >
                  Apply
                </Button>
              </ButtonGroup>
            </Box>
          </Box>
        </Box>
      </form>
    </Box>
  );
}

ChartConfigSelectorStep.propTypes = {
  existingConfig: PropTypes.shape({}),
  chartType: PropTypes.string.isRequired,
  onApply: PropTypes.func.isRequired,
  backToChartType: PropTypes.func.isRequired,
};

ChartConfigSelectorStep.defaultProps = {
  existingConfig: {},
};
export default ChartConfigSelectorStep;
