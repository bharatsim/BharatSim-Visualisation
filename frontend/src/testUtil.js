// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
