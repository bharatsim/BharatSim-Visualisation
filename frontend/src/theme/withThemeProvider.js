import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
