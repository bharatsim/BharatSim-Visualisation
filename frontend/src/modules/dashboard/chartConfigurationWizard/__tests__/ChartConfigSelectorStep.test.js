import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ChartConfigSelectorStep from '../ChartConfigSelectorStep';
import withThemeProvider from '../../../../theme/withThemeProvider';
import { withProjectLayout, withRouter } from '../../../../testUtil';

jest.mock('../../../../utils/api', () => ({
  api: {
    getDatasources: jest.fn().mockResolvedValue({
      dataSources: [
        { name: 'datasource1', _id: 'id1' },
        { name: 'datasource2', _id: 'id2' },
      ],
    }),
    getCsvHeaders: jest.fn().mockResolvedValue({
      headers: [
        { name: 'column1', type: 'number' },
        { name: 'column2', type: 'number' },
      ],
    }),
  },
}));
describe('chart config selector', () => {
  const ChartConfigSelectorStepWithTheme = withThemeProvider(
    withRouter(withProjectLayout(ChartConfigSelectorStep)),
  );
  it('should match snapshot', async () => {
    const { container, findByText } = render(
      <ChartConfigSelectorStepWithTheme
        chartType="lineChart"
        onApply={jest.fn()}
        backToChartType={jest.fn()}
      />,
    );
    await findByText('Data Source');
    expect(container).toMatchSnapshot();
  });
  

  it('should call back to chart type selector', async () => {
    const backToChartType = jest.fn();
    const { findByText, getByText } = render(
      <ChartConfigSelectorStepWithTheme
        chartType="lineChart"
        onApply={jest.fn()}
        backToChartType={backToChartType}
      />,
    );
    await findByText('Data Source');

    const backToChartTypeButton = getByText('Back to chart type').closest('button');
    fireEvent.click(backToChartTypeButton);

    expect(backToChartType).toHaveBeenCalled();
  });
});
