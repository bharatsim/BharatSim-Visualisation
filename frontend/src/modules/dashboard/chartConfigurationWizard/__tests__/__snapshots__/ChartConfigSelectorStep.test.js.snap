// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chart config selector should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <form>
      <div
        class="MuiBox-root MuiBox-root-8 makeStyles-tabContainer-3"
      >
        <div
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root makeStyles-tabRoot-4 MuiTab-textColorInherit Mui-selected"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Data
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root makeStyles-tabRoot-4 MuiTab-textColorInherit"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Styles
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-9 PrivateTabIndicator-colorPrimary-10 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-13 makeStyles-container-2"
      >
        <div
          class="MuiBox-root MuiBox-root-14"
        >
          <div
            class="MuiBox-root MuiBox-root-15"
          >
            <div
              class="MuiBox-root MuiBox-root-20 makeStyles-mainContainer-16"
            >
              <div
                class="MuiBox-root MuiBox-root-21"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2"
                >
                  Chart Name
                </h6>
              </div>
              <div
                class="MuiBox-root MuiBox-root-22 makeStyles-fieldContainer-18"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root makeStyles-input-25"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
                    data-shrink="true"
                  >
                    Enter chart name
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input"
                      data-testid="chart-name-input"
                      name="chartName"
                      type="text"
                      value="Untitled Chart"
                    />
                  </div>
                  
                </div>
              </div>
            </div>
          </div>
          <hr
            class="MuiDivider-root"
          />
          <div
            class="MuiBox-root MuiBox-root-26"
          >
            <div
              class="MuiBox-root MuiBox-root-67 makeStyles-mainContainer-16"
            >
              <div
                class="MuiBox-root MuiBox-root-68"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2"
                >
                  Data Source
                </h6>
              </div>
              <div
                class="MuiBox-root MuiBox-root-69 makeStyles-fieldContainer-18"
              >
                <div
                  class="MuiFormControl-root makeStyles-formControl-70"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                    data-shrink="false"
                    id="dropdown-label"
                  >
                    Select data source
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                    data-testid="dropdown-dataSources"
                  >
                    <div
                      aria-haspopup="listbox"
                      aria-labelledby="dropdown-label dropdown-dataSources"
                      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
                      id="dropdown-dataSources"
                      role="button"
                      tabindex="0"
                    >
                      <span>
                        â€‹
                      </span>
                    </div>
                    <input
                      aria-hidden="true"
                      class="MuiSelect-nativeInput"
                      tabindex="-1"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                  <p
                    class="MuiFormHelperText-root makeStyles-helperText-72 MuiFormHelperText-contained"
                  >
                    Data-source change may or may-not reset the chart configuration
                  </p>
                </div>
              </div>
            </div>
          </div>
          
        </div>
        <div
          class="MuiBox-root MuiBox-root-29 makeStyles-hide-6"
        >
          <div>
            <div
              class="MuiBox-root MuiBox-root-30"
            >
              <div
                class="MuiBox-root MuiBox-root-31"
              >
                <div
                  class="MuiBox-root MuiBox-root-35"
                >
                  <div
                    class="MuiBox-root MuiBox-root-36 makeStyles-textContainer-34"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-subtitle2"
                    >
                      Axis
                    </h6>
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-37 makeStyles-fieldsContainerVertical-32"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-38 makeStyles-mainContainer-16"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-39"
                      >
                        <h6
                          class="MuiTypography-root MuiTypography-subtitle2"
                        >
                          X Axis
                        </h6>
                      </div>
                      <div
                        class="MuiBox-root MuiBox-root-40 makeStyles-fieldContainer-18"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root makeStyles-input-25"
                        >
                          <label
                            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                            data-shrink="false"
                          >
                            Enter X axis label
                          </label>
                          <div
                            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiFilledInput-input"
                              data-testid="x-axis-title"
                              name="axisConfig.xAxisTitle"
                              type="text"
                              value=""
                            />
                          </div>
                          
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root MuiBox-root-41 makeStyles-mainContainer-16"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-42"
                      >
                        <h6
                          class="MuiTypography-root MuiTypography-subtitle2"
                        >
                          Y Axis
                        </h6>
                      </div>
                      <div
                        class="MuiBox-root MuiBox-root-43 makeStyles-fieldContainer-18"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root makeStyles-input-25"
                        >
                          <label
                            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                            data-shrink="false"
                          >
                            Enter Y axis label
                          </label>
                          <div
                            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiFilledInput-input"
                              data-testid="y-axis-title"
                              name="axisConfig.yAxisTitle"
                              type="text"
                              value=""
                            />
                          </div>
                          
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <hr
                class="MuiDivider-root"
              />
            </div>
            <div
              class="MuiBox-root MuiBox-root-44"
            >
              <div
                class="MuiBox-root MuiBox-root-45"
              >
                <div
                  class="MuiBox-root MuiBox-root-50 makeStyles-mainContainer-16"
                >
                  <div
                    class="MuiBox-root MuiBox-root-51"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-subtitle2"
                    >
                      Series
                    </h6>
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-52 makeStyles-fieldContainer-18"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-53 makeStyles-fieldArrayContainer-48"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-54 makeStyles-fallbackText-49"
                      >
                        <h6
                          class="MuiTypography-root MuiTypography-subtitle2"
                        >
                          Select Y axis measure to add styles
                        </h6>
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
              
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-55 makeStyles-footer-1"
        >
          <hr
            class="MuiDivider-root"
          />
          <div
            class="MuiBox-root MuiBox-root-56"
          >
            <div
              class="MuiBox-root MuiBox-root-58 makeStyles-buttonGroup-57"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Back to chart type
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Apply
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
